src/c_sources.ml

```diff
@@ -0,0 +1,130 @@
+open Stdune
+open Dune_file
+
+module Library = Dune_file.Library
+
+type t =
+  { libraries : C.Sources.t Lib_name.Map.t
+  }
+
+let for_lib t ~dir ~name =
+  match Lib_name.Map.find t.libraries name with
+  | Some m -> m
+  | None ->
+    Exn.code_error "C_sources.for_lib"
+      [ "name", Lib_name.to_sexp name
+      ; "dir", Path.to_sexp dir
+      ; "available", Sexp.Encoder.(list Lib_name.to_sexp)
+                       (Lib_name.Map.keys t.libraries)
+      ]
+
+let empty =
+  { libraries = Lib_name.Map.empty
+  }
+
+let c_name, cxx_name =
+  let make what ~loc s =
+    if match s with
+      | "" | "." | ".."  -> true
+      | _ -> false then
+      Errors.fail loc "%S is not a valid %s name." s what
+    else
+      s
+  in
+  ( make "C"
+  , make "C++"
+  )
+
+module Eval = struct
+  module Value = struct
+    type t = string
+    type key = string
+    let key s = s
+  end
+
+  include Ordered_set_lang.Make_loc(String)(Value)
+end
+
+let load_sources ~dir ~files =
+  let init = C.Kind.Dict.make String.Map.empty in
+  String.Set.fold files ~init ~f:(fun fn acc ->
+    match C.Kind.split_extension fn with
+    | None -> acc
+    | Some (obj, kind) ->
+      let path = Path.relative dir fn in
+      C.Kind.Dict.update acc kind ~f:(fun v ->
+        String.Map.add v obj (C.Source.make ~kind ~path)
+      ))
+
+let make (d : _ Dir_with_dune.t)
+      ~(c_sources : C.Source.t String.Map.t C.Kind.Dict.t) =
+  let libs =
+    List.filter_map d.data ~f:(fun stanza ->
+      match (stanza : Stanza.t) with
+      | Library lib ->
+        let eval (kind : C.Kind.t) (c_sources : C.Source.t String.Map.t)
+              validate osl =
+          Eval.eval_unordered osl
+            ~parse:(fun ~loc s ->
+              let s = validate ~loc s in
+              let s' = Filename.basename s in
+              if s' <> s then begin
+                Errors.warn loc "relative part of stub are no longer \
+                                 necessary and are ignored."
+              end;
+              s'
+            )
+            ~standard:String.Map.empty
+          |> String.Map.map ~f:(fun (loc, s) ->
+            match String.Map.find c_sources s with
+            | Some source -> (loc, source)
+            | None ->
+              Errors.fail loc "%s does not exist as a C source. \
+                               One of %s must be present"
```

Comment:
Nitpicking, but when there is a single element in `C.Kind.possible_fns kind s`, the error message won't be optimal:

```
file does not exist as a C source.
One of file.c must be present
```

it would read nicer as:

```
file does not exist as a C source.
file.c must be present
```